apps/dhparam.c

```diff
@@ -191,221 +176,237 @@ int dhparam_main(int argc, char **argv)
                 goto end;
             }
-            if (!EVP_PKEY_paramgen(ctx, &pkey)) {
-                BIO_printf(bio_err, "Error, DH generation failed\n");
+            if (!EVP_PKEY_CTX_set_dh_paramgen_generator(ctx, g)) {
+                BIO_printf(bio_err, "Error, unable to set generator\n");
                 goto end;
             }
-            dh = EVP_PKEY_get0_DH(pkey);
         }
+
+        if (!EVP_PKEY_paramgen(ctx, &tmppkey)) {
+            BIO_printf(bio_err, "Error, %s generation failed\n", alg);
+            goto end;
+        }
+
+        EVP_PKEY_CTX_free(ctx);
+        ctx = NULL;
+        if (dsaparam) {
+            pkey = dsa_to_dh(tmppkey);
+            if (pkey == NULL)
+                goto end;
+            EVP_PKEY_free(tmppkey);
+        } else {
+            pkey = tmppkey;
+        }
+        tmppkey = NULL;
     } else {
+        OSSL_DECODER_CTX *decoderctx = NULL;
+        const char *keytype = "DH";
+        int done;
+
         in = bio_open_default(infile, 'r', informat);
         if (in == NULL)
             goto end;
 
-#if !defined(OPENSSL_NO_DSA) && !defined(OPENSSL_NO_DEPRECATED_3_0)
-        if (dsaparam) {
-            DSA *dsa;
-
-            if (informat == FORMAT_ASN1)
-                dsa = d2i_DSAparams_bio(in, NULL);
-            else                /* informat == FORMAT_PEM */
-                dsa = PEM_read_bio_DSAparams(in, NULL, NULL, NULL);
+        do {
+            /*
+             * We assume we're done unless we explicitly want to retry and set
+             * this to 0 below.
+             */
+            done = 1;
+            /*
+            * We set NULL for the keytype to allow any key type. We don't know
+            * if we're going to get DH or DHX (or DSA in the event of dsaparam).
+            * We check that we got one of those key types afterwards.
+            */
+            decoderctx
+                = OSSL_DECODER_CTX_new_by_EVP_PKEY(&tmppkey,
+                                                    (informat == FORMAT_ASN1)
+                                                    ? "DER" : "PEM",
+                                                    (informat == FORMAT_ASN1)
+                                                    ? keytype : NULL,
+                                                    NULL, NULL);
+
+            if (decoderctx != NULL
+                    && !OSSL_DECODER_from_bio(decoderctx, in)
+                    && informat == FORMAT_ASN1
+                    && strcmp(keytype, "DH") == 0) {
+                /*
+                * When reading DER we explicitly state the expected keytype
+                * because, unlike PEM, there is no header to declare what
+                * the contents of the DER file are. The decoders just try
+                * and guess. Unfortunately with DHX key types they may guess
+                * wrong and think we have a DSA keytype. Therefore we try
+                * both DH and DHX sequentially.
```

Comment:
Hmmm...  OSSL_STORE in 1.1.1 has functionality that checks for such ambiguity, i.e. it tries all possible decodings and keeps track of how many were successful, and returns an ambiguity error when they are more than one.  That check was never transported to OSSL_DECODER...  it rather just takes the first success and returns that, with no further decoding attempts.